### ISSUE TITLE
Start from the result of the last key lookup when doing the next key lookup

### ISSUE THREAD

**Issue by michaeleisel** [simdjson/simdjson#677]
Typically, the order of key lookups for ZippyJSON matches the order of the keys. So if it starts from the last lookup, then processing the object can go from O(n^2) to O(n)

**Comment by lemire** [comment]
The document-object-model in simdjson is immutable. It does not keep track of how it was accessed.

**Comment by michaeleisel** [comment]
Maybe it could just *appear* immutable?

**Comment by michaeleisel** [comment]
Or perhaps a hint could be returned for future searches, so that the consumer manages the state

**Comment by jkeiser** [comment]
Yeah, it's entirely possible to modify `dom::object` to have an extra `last_position` or something. If we did the same for arrays, index access could even become O(1) for all the common cases ...

**Comment by lemire** [comment]
If we do away with immutability, we should document the concurrency issues that will arise:

https://github.com/simdjson/simdjson/issues/681

**Comment by michaeleisel** [comment]
What if it returns an iterator, that can be passed to an overload of the lookup function

**Comment by jkeiser** [comment]
Very little of simdjson is thread safe anyway, we include a note about it. Dom references are currently thread safe as long as no other thread calls parse; if that happens, the document will be overwritten and the dom references become garbage.

Iterators are particularly cheap to copy, too I would not feel bad if they were not thread safe.

But we have time to talk it all through :)

**Comment by michaeleisel** [comment]
fyi i ended up using an iterator to do this, as such. `iteratorPtr` gets updated at the end of each call to `JNTDocumentFindValue`, so that future calls to the function can start there

```
simdjson_result<dom::element> JNTDocumentFindValue(JNTDecoder decoder, const char *cKey, dom::object::iterator *iteratorPtr) {
    auto iterator = *iteratorPtr;
    std::string_view key = cKey;
    const auto &searchStart = iterator;
    const dom::object &object = decoder.element;
    const auto &end = object.end();
    dom::element child;
    bool found = false;
    while (iterator != end) {
        if (key == iterator.key()) {
            child = iterator.value();
            found = true;
            break;
        }
        ++iterator;
    }
    if (!found) {
        iterator = object.begin();
        while (iterator != searchStart) {
            if (key == iterator.key()) {
                child = iterator.value();
                found = true;
                break;
            }
            ++iterator;
        }
    }
    if (!found) {
        return simdjson_result<dom::element>(NO_SUCH_FIELD);
    }
    *iteratorPtr = iterator;
    return simdjson_result<dom::element>(std::move(child));
}
```

**Comment by lemire** [comment]
The On Demand API does exactly this. `ondemand::object::find_field_unordered()` (and `operator[]`, which calls it) resumes where the previous lookup stopped and wraps around to the start of the object only if the key isn't found, so looking up keys in document order is linear overall. `find_field()` is the strictly forward-only variant. The DOM is immutable by design, and the iterator approach shown above still works there. Closing.
